### PR TITLE
Extract ICMP echo responders to their own apps

### PR DESCRIPTION
**Pull request by wingo**
This PR extracts the remaining general-purpose IP apps from the lwaftr.  It also converts them to FFI style instead of offset style.

Although I think FFI style is better from a reliability perspective -- there's less likelihood of making errors because the FFI types compute their offsets automatically; there's less casting happening -- there is more boilerplate duplication at the top of files.  We could solve this by using the `lib.protocol` family of libraries.  That would also have the advantage of hiding the various byte swapping and bitfield accessing ceremony behind methods, which would be really great if the impedance mismatch goes away.  It would also allow more utility functionality to be implemented by methods.  However I still find the lifecycle management of `lib.protocol` and especially of datagrams to be a bit too magical for my tastes.  Dunno.  I wonder if there is room for a simple IPv4 / IPv6 / Ethernet / ICMP / TCP / UDP library that's just FFI data types with a metatable.  Or is that precisely what the `lib.protocol` libraries are and I'm not seeing it?

In any case I think the style that we had settled on that uses raw `packet.data`, offsets from a `lwaftr.constants` module, and random utility functions wasn't a good style.  What do you think @dpino and @xray7224 ?

### CODE REVIEW

**Review by tsyesika — APPROVED** [review]
LGTM. Huge readability improvement to the old code 👍

**Comment by dpino** [comment]
LGTM. I have one concern though and it's whether it works if a packet has VLAN tags. The ethernet template is for plain ethernet but 802.1Q has the inconvenient of adding the extra vlan field between source mac address and ethertype. AFAIK, the old code worked with VLAN tags because the ethernet_offset in constants was adjusted globally. Anyway, there are end-to-end tests with VLAN tags, if the test passes that means it works (although I don't understand how they would).

As for using Alex Gall protocol libraries, that could be a possibility. As you know we discard them for the lwAFTR because they added overhead and it was faster to manipulate the packets directly. But for these libraries they could be an option as they are not as performance demanding.

The libraries are a bit magical that's true. They are architectured following a class hierarchy. Base class is called `header` I think. They use ffi templates as well. To build specific headers out of an existing packet there's a method called `new_from_mem`. I guess to really know if they are worth using is to compare performance using this approach and using the protocol libraries. Would require to rewrite the app in the protocol style. I'm OK with this straight ffi approach. Once the apps have landed upstream perhaps with the help of Alex Gall can give them a new spin.

File: src/apps/ipv4/echo.lua [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+end
+
+function ICMPEcho:respond_to_echo_request(pkt)
+   -- Pass on packets too small to be ICMPv4.
+   local min_len = ether_header_len + min_ipv4_header_len + icmp_header_len
+   if pkt.length < min_len then return false end
+
+   -- Is it ICMPv4?
+   local h = ffi.cast(ether_ipv4_header_ptr_t, pkt.data)
+   if ntohs(h.ether.type) ~= ether_type_ipv4 then return false end
```
Comment (dpino) [inline src/apps/ipv4/echo.lua]: This check works either for Ethernet headers and VLAN headers. So my doubt is solved.

**Comment by wingo** [comment]
I think it used to be the case that the lwaftr would globally adjust a constant indicating the offset of the L3 header for VLAN-related reasons, but that hasn't been the case for a long time.  At some point we switched to stripping VLAN tags on ingress, either with hardware facilities where available, or via the Untagger vlan app.  Since then, all offsets have been absolute.  And indeed often we don't even allow IPv6 extension headers!  That should probably be fixed at some point, but anyway to my knowledge the lwAFTR no longer depends on global state in that way.  This does allow us to instantiate multiple lwAFTR instances with different configurations in the same app graph, even if vlan settings conflict, though this isn't a capability that we use or even are planning on using...

**Comment by dpino** [comment]
Thanks for the reminder, it has helped me to clarify the whole picture (I was confused about this). Now I remember that the end-to-end tests (`lwaftr check` actually) use the VLAN app to untag/tag packets and that's why it works, packets reach the Echo app with a standard Etthernet header.

**Pull request merged**